### PR TITLE
D2k - Limit Production to Atreides and Ordos Palaces

### LINKED ISSUE #14660: Fremen spawning from the wrong palace
From https://logs.openra.net/?year=2018&month=01&day=04#02:26:53

>[02:25:52] | < muh> Hi
>[02:26:53] | < muh>  I noticed that if you capture both the atreides and the corinno palace  in Harkonnen 9a (and 9b I assume) fremen support power spawns at the  corinno palace instead of atreides
>[02:27:08] | < muh> I dont know if that is expected behaviour but I guess not

### PR DESCRIPTION
**Pull request by MustaphaTR**
Fixes #14660 and some other palace related issues.

Last commit is for testing but we may keep it if we want.

### CODE REVIEW

**Comment by Arular101** [comment]
Please don't forget to update the copyright notice year.

**Comment by MustaphaTR** [comment]
Updated the year.

**Comment by ltem** [comment]
`PrimaryBuilding.cs` and `SupportPowerChargeBar.cs` still have the old `2007-2017` copyright.
- While

**Comment by MustaphaTR** [comment]
That's because i haven't rebased. It shouldn't be a problem.

**Comment by ltem** [comment]
Ahh sorry, I misclicked...

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnFaction.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	{
+		[FieldLoader.Require]
+		[GrantedConditionReference]
+		[Desc("Condition to grant.")]
+		public readonly string Condition = null;
+
+		[Desc("Only grant this condition for certain factions.")]
+		public readonly HashSet<string> Factions = new HashSet<string>();
+
+		[Desc("Should it recheck everything when it is captured?")]
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnFaction.cs]: Am no native speaker, what about `Reevaluate if the condition should still be granted when this actor's owner changes.` though?

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnFaction.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+		{
+			conditionManager = self.Trait<ConditionManager>();
+
+			base.Created(self);
+		}
+
+		public void OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
+		{
+			if (Info.ResetOnOwnerChange)
+				faction = newOwner.Faction.InternalName;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnFaction.cs]: Is there some trait enabling/disabling going on while the owner changes or do we still need to call the rechecking from here manually?

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnFaction.cs (hunk 1 above)
Comment (MustaphaTR) [reply OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnFaction.cs]: I have copy pasted that from ProvidesPrerequisite, i think it is better to keep that for consistency, or i can edit both.

**Review by abcdefg30 — COMMENTED** [review]
Death Hand missiles can still spawn from all palaces. You can also build Fremen at an Ordos palace and Saboteurs at an Atreides one.

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnFaction.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+		protected override void Created(Actor self)
+		{
+			conditionManager = self.Trait<ConditionManager>();
+
+			base.Created(self);
+		}
+
+		public void OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
+		{
+			if (Info.ResetOnOwnerChange)
```
[the remark this answers is not in the document]
Comment (MustaphaTR) [reply OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnFaction.cs]: I don't think ProvidesPrerequisite has such check, should i add there too while i'm in it?

**Comment by MustaphaTR** [comment]
PrimaryBuilding wasn't happy with multipile Produces so i couldn't seperate Saboteur and Fremen. Didn't really get into to code to fix, will take a look what i can do.

I tryed to fix the Death Hand, but couldn't find where the activator building is selected. It looks like it is nearest one unless AllowMultipile is true to target and i think it needs a IsTraitDisabled check.

**Comment by pchote** [comment]
I didn't test it, but something like the following might help:
```diff
diff --git a/OpenRA.Mods.Common/Traits/Buildings/PrimaryBuilding.cs b/OpenRA.Mods.Common/Traits/Buildings/PrimaryBuilding.cs
index 4c0159cd9f..e13ed41ca0 100644
--- a/OpenRA.Mods.Common/Traits/Buildings/PrimaryBuilding.cs
+++ b/OpenRA.Mods.Common/Traits/Buildings/PrimaryBuilding.cs
@@ -90,7 +90,7 @@ public void SetPrimaryProducer(Actor self, bool isPrimary)
                        {
                                // Cancel existing primaries
                                // TODO: THIS IS SHIT
-                               var queues = info.ProductionQueues.Length == 0 ? self.Info.TraitInfo<ProductionInfo>().Produces : info.ProductionQueues;
+                               var queues = info.ProductionQueues.Length == 0 ? self.Info.TraitInfos<ProductionInfo>().SelectMany(pi => pi.Produces) : info.ProductionQueues;
                                foreach (var q in queues)
                                {
                                        foreach (var b in self.World
@@ -99,7 +99,7 @@ public void SetPrimaryProducer(Actor self, bool isPrimary)
                                                                a.Actor != self &&
                                                                a.Actor.Owner == self.Owner &&
                                                                a.Trait.IsPrimary &&
-                                                               a.Actor.Info.TraitInfo<ProductionInfo>().Produces.Contains(q)))
+                                                               a.Actor.Info.TraitInfos<ProductionInfo>().Any(pi => pi.Produces.Contains(q))))
                                                b.Trait.SetPrimaryProducer(b.Actor, false);
                                }
 
```

**Comment by pchote** [comment]
RE support powers: There are several places in `SupportPowerManager` that checks if the trait is paused, not not if it is disabled.  That should be an easy fix.

**Comment by MustaphaTR** [comment]
Updated.

**Comment by pchote** [comment]
Ordos and Atreides palaces will override eachothers primary status, but produce different units.  If we already support multiple `PrimaryBuilding`s on an actor then they should be split, otherwise I don't think this should block merging.

File: OpenRA.Mods.Common/Traits/Render/SupportPowerChargeBar.cs [hunk 4]
```diff
@@ -16,43 +16,45 @@
...
-			this.info = info;
 			spm = self.Owner.PlayerActor.Trait<SupportPowerManager>();
 		}
 
 		float ISelectionBar.GetValue()
 		{
+			if (IsTraitDisabled)
+				return 0;
+
 			var power = spm.GetPowersForActor(self).FirstOrDefault(sp => !sp.Disabled);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/SupportPowerChargeBar.cs]: This can return powers that aren't available on the specific actor.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Render/SupportPowerChargeBar.cs]: Fix with
```diff
diff --git a/OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs b/OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs
index 129ddcb3bf..ae34031d2e 100644
--- a/OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs
+++ b/OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs
@@ -124,7 +124,8 @@ public IEnumerable<SupportPowerInstance> GetPowersForActor(Actor a)
                                return NoInstances;
 
                        return a.TraitsImplementing<SupportPower>()
-                               .Select(t => Powers[MakeKey(t)]);
+                               .Select(t => Powers[MakeKey(t)])
+                               .Where(p => p.Instances.Any(i => !i.IsTraitDisabled && i.Self == a));
                }
 
                public void PrerequisitesAvailable(string key)
```

File: OpenRA.Mods.Common/Traits/Buildings/PrimaryBuilding.cs [hunk 5]
```diff
@@ -99,17 +102,25 @@ public void SetPrimaryProducer(Actor self, bool isPrimary)
 								a.Actor != self &&
 								a.Actor.Owner == self.Owner &&
 								a.Trait.IsPrimary &&
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/PrimaryBuilding.cs]: Adding `a.Trait.Info.ProductionQueues.Intersect(Info.ProductionQueues).Any()` here should fix the issue stopping you from spliting PrimaryBuilding status.

**Review by reaperrr — APPROVED** [review]
Looks good to me 👍

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(Next-Playtest)/_compare/62364548a2fe4c0c0e35cb13f1c2a90a83e3b052...224645b479b6f2d9a51faf2fb369875e0f4d9d81)

https://github.com/OpenRA/OpenRA/commit/645b181af767e1ee1868134d6488792cdd548bde
https://github.com/OpenRA/OpenRA/commit/766c091c1ca909626ff72ca2a35502223893bfb5
https://github.com/OpenRA/OpenRA/commit/94028c8246bb4e4faadce32f25f9bc8d46fa587d
https://github.com/OpenRA/OpenRA/commit/d110b9c648dd41abce004a0ac412366e0909693c
https://github.com/OpenRA/OpenRA/commit/5b0875f66f3ddd088301a79d0b6bd21342b96a0f
https://github.com/OpenRA/OpenRA/commit/bdaff276566374c8df28c63c2c28fa42220c80b9

**Pull request merged**